### ISSUE TITLE
Unexpected error when calling Codec(En|De)codeSelf directly, bypassing enc.Encode/dec.Decode

### ISSUE THREAD

**Issue by shory152** [ugorji/go#199]
decoding an array into a slice, report error. maybe the bd is invalid  at func (d *msgpackDecDriver) ContainerType()

**Comment by ugorji** [comment]
Can you please detail what you were trying to do, what you expected, and what you saw, and attach a reproducer that I can run using "go run file.go" or "go test"

Right now, I can't do anything with this.

**Comment by shory152** [comment]
Please refer to the test case at #200

**Comment by ugorji** [comment]
Please State what you are trying to do, what you expect to happen, what happens, and how to reproduce it. Right now, you just gave me code, and I don't know what to do with it. 

Time is tight. I depend on you being clear.

Also, please put the information in #200 in this issue, and close that one. I shouldn't be jumping around to debug same issue

**Comment by shory152** [comment]
msg_test.go:
```
package shory

import "reflect"
import "github.com/ugorji/go/codec"
import "testing"
import "bytes"

func TestMsg(t *testing.T){

	var mh codec.MsgpackHandle
	var h *codec.MsgpackHandle
	
	mh.MapType = reflect.TypeOf(map[string]interface{}(nil))
	mh.SliceType = reflect.TypeOf([]interface{}(nil))
	mh.EncodeOptions.StructToArray = true
	h = &mh
	
        // encode msg1 to package
	msg1 := Msg{ID:1,Name:"msg1",Value:"value1"}

	pkgbuf := new(bytes.Buffer)
	enc := codec.NewEncoder(pkgbuf,h)

	msg1.CodecEncodeSelf(enc)
	
        // decode package to msg2
	dec := codec.NewDecoder(pkgbuf,h)
	var msg2 Msg
	msg2.CodecDecodeSelf(dec) // NOTE: report error "only slice or array can be decode to array"
	
	if msg1.ID == msg2.ID && msg1.Name == msg2.Name && msg1.Value == msg2.Value {
		t.Log("decode msg OK:", msg2)
	} else {
		t.Error("decode msg ERR:", msg2)
	}
}

```

msg.go:
```
package shory

import "github.com/ugorji/go/codec"

type Msg struct {
	ID int
	Name string
	Value string
}

type MsgSlice []Msg

type Dummy struct{}
func (d *Dummy)CodecEncodeSelf(encoder *codec.Encoder){}
func (d *Dummy)CodecDecodeSelf(decoder *codec.Decoder){}
```

codecgen -o msg.gen.go msg.go

**Comment by ugorji** [comment]
You have attached a reproducer, but I don't know what I am reproducing. 

For next time, please fill in the following sections in a comment below. I will put some sample responses. This ensures that I don't have to infer what you are trying to do, and spin cycles trying to understand why it is not working, and you end up telling me that your issue was with something else. That's why we always ask that you explicitly state it.

**What you are trying to do?**
I really don't know what you were trying to do? It can be as simple as: I am trying to encode a custom type MsgSlice []Msg in the sample below using codegen. However, I want to directly call CodecEncodeSelf, as opposed to using enc.Encode directly.

**what do you expect to happen?**
I expect that calling CodecEncodeSelf works just as calling enc.Encode

**what happens?**
enc.Encode works as expected. However, when I directly call msg1.CodecEncodeSelf, it fails

**How can I reproduce it?**
Copy the files to an empty folder, run codecgen -o msg.gen.go msg.go, and then run "go test".

**Comment by ugorji** [comment]
Seems the issue is that you are calling "CodecEncodeSelf" directly.

In general, these calls for Codec(En|de)codeSelf should not be called directly. The enc.Encode and dec.Decode do some pre and post setup before calling these, and you miss those when calling directly.

You should just call enc.Encode or dec.Decode calls instead. That will fix the issues you are seeing. I replaced those calls with these in your code, and it works fine. I also put multiple println calls in the generated code, and confirmed that it is hitting your generated code. Feel free to verify.

**Comment by shory152** [comment]
Thank you very much!

I modified msgpack.go:571 like below:
```
func (d *msgpackDecDriver) ContainerType() (vt valueType) {
    if !d.bdRead {
        d.readNextBd()
    }
...
```
then calling CodecDecodeSelf() is OK.

**calling CodecDecodeSelf is faster than calling dec.Decode()**

**Comment by ugorji** [comment]
Nice catch.

**Comment by ugorji** [comment]
Fixed with 5efa325
